File: AndroidTai/app/src/main/java/com/tai/androidtai/AndroidTaiApplication.java
package com.tai.androidtai;

import android.app.Activity;
import android.app.Application;

import com.facebook.drawee.backends.pipeline.Fresco;
import com.tai.androidtai.dagger.components.AndroidTaiComponent;
import com.tai.androidtai.dagger.components.ApplicationComponent;

import javax.inject.Inject;

import androidx.annotation.VisibleForTesting;
import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.support.DaggerApplication;

public class AndroidTaiApplication extends DaggerApplication {

    @Inject
    @VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
    public DispatchingAndroidInjector<Activity> mActivityDispatchingAndroidInjector;

    AndroidTaiComponent mApplicationComponent;

    @Override
    public void onCreate() {
        super.onCreate();
        Fresco.initialize(this);
        mApplicationComponent = initializeApplicationComponent();
        mApplicationComponent.inject(this);
        Fresco.initialize(this);
    }

    AndroidTaiComponent initializeApplicationComponent() {
        return ApplicationComponent.Initializer.Companion.init(this);
    }

    public AndroidTaiComponent getApplicationComponent() {
        return mApplicationComponent;
    }

    @Override
    protected AndroidInjector<? extends DaggerApplication> applicationInjector() {
        mApplicationComponent = ApplicationComponent.Initializer.Companion.init(this);
        mApplicationComponent.inject(this);
        return mApplicationComponent;
    }

}

File: AndroidTai/app/src/main/java/com/tai/androidtai/AndroidTaiApplication.kt
package com.tai.androidtai

import android.app.Activity
import androidx.annotation.VisibleForTesting
import com.facebook.drawee.backends.pipeline.Fresco
import com.tai.androidtai.dagger.components.AndroidTaiComponent
import com.tai.androidtai.dagger.components.ApplicationComponent
import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.support.DaggerApplication
import javax.inject.Inject

class AndroidTaiApplication : DaggerApplication() {

    @Inject
    @VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
    lateinit var mActivityDispatchingAndroidInjector: DispatchingAndroidInjector<Activity>

    lateinit var applicationComponent: AndroidTaiComponent

    override fun onCreate() {
        super.onCreate()
        Fresco.initialize(this)
        applicationComponent = initializeApplicationComponent()
        applicationComponent.inject(this)
        Fresco.initialize(this)
    }

    private fun initializeApplicationComponent(): AndroidTaiComponent {
        return ApplicationComponent.Initializer.init(this)
    }

    override fun applicationInjector(): AndroidInjector<out DaggerApplication> {
        applicationComponent = ApplicationComponent.Initializer.init(this)
        applicationComponent.inject(this)
        return applicationComponent
    }

}